core/resharding.go

```diff
@@ -94,25 +94,35 @@ func (ss *ShardingState) cuckooResharding(percent float64) {
 	}
 }
 
-// assignLeaders will first add new nodes into shards, then use cuckoo rule to reshard to get new shard state
-func (ss *ShardingState) assignLeaders() {
+// Reshard will first add new nodes into shards, then use cuckoo rule to reshard to get new shard state
+func (ss *ShardingState) Reshard(newNodeList []types.NodeID, percent float64) {
+	rand.Seed(int64(ss.rnd))
+	ss.sortCommitteeBySize()
+
+	// Take out and preserve leaders
+	leaders := []types.NodeID{}
 	for i := 0; i < ss.numShards; i++ {
-		// At genesis epoch, the shards are empty.
 		if len(ss.shardState[i].NodeList) > 0 {
+			leaders = append(leaders, ss.shardState[i].NodeList[0])
+			utils.GetLogInstance().Info("LEADER===01", "shard", ss.shardState[i])
+			ss.shardState[i].NodeList = ss.shardState[i].NodeList[1:]
+			utils.GetLogInstance().Info("LEADER===02", "shard", ss.shardState[i])
+			// Also shuffle the rest of the nodes
 			Shuffle(ss.shardState[i].NodeList)
-			ss.shardState[i].Leader = ss.shardState[i].NodeList[0]
+			utils.GetLogInstance().Info("LEADER===03", "shard", ss.shardState[i])
 		}
 	}
-}
 
-// Reshard will first add new nodes into shards, then use cuckoo rule to reshard to get new shard state
-func (ss *ShardingState) Reshard(newNodeList []types.NodeID, percent float64) {
-	rand.Seed(int64(ss.rnd))
-	ss.sortCommitteeBySize()
-	// TODO: separate shuffling and leader assignment
-	ss.assignLeaders()
+	utils.GetLogInstance().Info("LEADER===1", "leaders", leaders)
 	ss.assignNewNodes(newNodeList)
 	ss.cuckooResharding(percent)
+
+	// Put leader back
+	for i := 0; i < ss.numShards; i++ {
+		ss.shardState[i].NodeList = append(leaders[i:i+1], ss.shardState[i].NodeList...)
```

Comment:
please do index out of range check here.

Reply:
Done.